cfme/base/credential.py

```diff
@@ -0,0 +1,224 @@
+# -*- coding: utf-8 -*-
+from cfme.web_ui import FileInput, Input, Radio, form_buttons
+from cfme.web_ui.tabstrip import TabStripForm
+from utils import conf
+from utils.pretty import Pretty
+from utils.update import Updateable
+
+
+class FromConfigMixin(object):
+    @classmethod
+    def from_config(cls, key):
+        creds = conf.credentials[key]
+
+        # todo: fix this along with other credential fixes. code or credentials conf ?
+        to_rename = [('password', 'secret'), ('username', 'principal')]
+        for key1, key2 in to_rename:
+            if key1 in creds:
+                creds[key2] = creds[key1]
+                del creds[key1]
+
+        return cls(**creds)
+
+
+class Credential(Pretty, Updateable, FromConfigMixin):
+    """
+    A class to fill in credentials
+
+    Args:
+        principal: user name
+        secret: password
+        verify_secret: password
+        domain: concatenated with principal if defined
+    """
+    pretty_attrs = ['principal', 'secret']
+
+    def __init__(self, principal, secret, verify_secret=None, domain=None, **ignore):
+        self.principal = principal
+        self.secret = secret
+        self.verify_secret = verify_secret
+        self.domain = domain
+
+    def __getattribute__(self, attr):
+        if attr == 'verify_secret':
+            if object.__getattribute__(self, 'verify_secret') is None:
+                return object.__getattribute__(self, 'secret')
+            else:
+                return object.__getattribute__(self, 'verify_secret')
+
+        elif attr == 'principal':
+            domain = object.__getattribute__(self, 'domain')
+            principal = object.__getattribute__(self, 'principal')
```

Comment:
this is an "evil attribute" we must mark that as tech debt to really remove __soon__